design/conditional-status-informers.md

```diff
@@ -0,0 +1,36 @@
+# Conditional Status Informers
+
+Currently, KOTS does not support having conditional application status informers.
+This proposal makes it possible to have a status informer disappear/appear if a specific config option is set to a certain value.
+
+## Goals
+
+- Add the ability to exclude specific status informers based on the applications's config values
+
+## Non Goals
+
+- Make the application status informers support additional kinds/objects.
+
+## Background
+
+For example, many vendors have an embedded vs. external database option.
+They have a status informer set up on the database object, but this only works when using the embedded option.
+Many of them have requested the ability to have conditional status informers so that the status informer can be excluded when selecting the external db option.
+
+## High-Level Design
+
+Add the ability to render/template the `statusInformers` array in the kots application spec (`kind: Application`).
+This way, config template functions (e.g. ConfigOptionEquals) can be used to decide whether or not an informer should be included/applied.
+
+## Detailed Design
+
+## Notes:
+- Getting the list of status informers happens during the app deploy loop in TypeScript.
+- The list of informers is then sent through a socket to the operator to be applied.
+
+1- To add the ability to exclude specific status informers, the `statusInformers` array will be rendered using the `RenderFile` function from kots ffi before it's sent to the operator (the `RenderFile` function can take any string as input to render).
+2- This way, conditional template functions can be used as an entry in the `statusInformers` array in the kots application spec.
+3- The template function will then resolve to either a valid status informer entry, or not (an empty string `""` for example).
+4- Invalid status informers (entries) will be excluded from the array by the operator.
```

Comment:
Do we currently silently ignore invalid status informers? This seems risky. If someone typos something, we shouldn't silently ignore it. This seems like the intent is to provide an alternate way to ignore informers for missing objects, but since we provide a way to explicitly handle that, I'd rather than also have an implicit way to handle it, that could backfire.

Reply:
Currently, if a status informer fails to parse (has invalid format), we only log the error. In some cases, this might lead to the app status being shown as `Missing`, but I think we need to improve the UX for this.

Reply:
Yeah, some improvements on the UX would be good there. Any suggestions?

Reply:
If a status informer is not running, the application status will not be `Ready`, in this case (or maybe in all cases), we can show a list of the running status informers along with the ones that are not running. This can help the user detect typos in the status informers. Also, we can update the linter to look for those types of mistakes, where it can detect if a status informer is pointing to a non-existing object (even when using template functions).

Reply:
@salahalsaleh what is the difference between an invalid informer and an empty string informer?

Also, can you link to code where these operations will happen?  As far as I know there is no way to render anything in TS.   Are we adding new FFI functions for this?

Reply:
@divolgin An invalid informer will typically have a typo, but an empty string informer will be the result of templating and be excluded.

We do have an existing ffi render function in TS and it's used in multiple places: https://github.com/replicatedhq/kots/blob/ca6a40ba/kotsadm/api/src/kots_app/kots_ffi.ts#L202